<launch>

    <arg name="namespace" default="sam"/>
    <arg name="mode" default="default"/>
    <arg name="debug" default="0"/>

    <arg name="x" default="0"/>
    <arg name="y" default="0"/>
    <arg name="z" default="0"/>
    <arg name="roll" default="0.0"/>
    <arg name="pitch" default="0.0"/>
    <arg name="yaw" default="0.0"/>

    <!-- STIM300 IMU driver -->
    <node pkg="stim300_ros_driver" name="stim_imu_driver" type="driver">
        <param name="device_path" value="/dev/ttyUSB0" />
        <param name="imu_frame" value="$(arg namespace)/imu_link" />
        <!-- <param name="imu_frame" value="$(arg namespace)/base_link" /> -->
        <param name="imu_output" value="/stim_imu" />
    </node>

    <!-- Depth from pressure sensor converter -->
    <node pkg="sam_dead_reckoning" name="press_to_depth" type="press_to_depth.py" output="screen">
        <param name="odom_frame" value="$(arg namespace)_odom" />
        <param name="pressure_topic" value="/uavcan_pressure2" />
        <param name="depth_topic" value="/uavcan_depth2" />
    </node>

    <node pkg="sam_dead_reckoning" name="sam_dr" type="sam_dr.py" output="screen">
        <param name="thrust_dr" value="$/sam_thrust_dr" />
        <param name="thrust_fb" value="/uavcan_to_ros_bridge_node/esc_status" />
        <!-- <param name="depth_topic" value="/uavcan_depth2" /> -->
    </node>

    <!-- EKF for fusing continuous position data from wheel odometry and IMU -->
    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_loc_local" output="screen" clear_params="true">
        <param name="frequency" value="30"/>
        <param name="sensor_timeout" value="0.1"/>
        <param name="two_d_mode" value="false"/>

        <!-- <param name="map_frame" value="map"/> -->
        <param name="odom_frame" value="$(arg namespace)_odom"/>
        <param name="base_link_frame" value="$(arg namespace)/base_link"/>
        <param name="world_frame" value="$(arg namespace)_odom"/>

        <param name="transform_time_offset" value="0.0"/>
        <param name="transform_timeout" value="0.0"/>

        <!--
            x,     y,      z,
            roll,  pitch,  yaw,
            vx,    vy,     vz,
            vroll, vpitch, vyaw,
            ax,    ay,     az
         -->

        <rosparam param="initial_state">[$(arg x), $(arg y), $(arg z),
                                         $(arg roll), $(arg pitch), $(arg yaw),
                                         0.0,  0.0,  0.0,
                                         0.0,  0.0,  0.0,
                                         0.0,  0.0,  0.0]</rosparam>

        <!-- Depth sensor -->
        <param name="pose0" value="/uavcan_depth2"/>
        <rosparam param="pose0_config">[true, true, true,
                                        false, false, false,
                                        false, false, false,
                                        false, false, false,
                                        false, false, false]</rosparam>

        <param name="pose0_differential" value="false"/>
        <param name="pose0_relative" value="false"/>


        <!-- IMU -->
        <param name="imu0" value="/stim_imu"/>
        <rosparam param="imu0_config">[false, false, false,
                                       true, true, false,
                                       false, false, false,
                                       true, true, true,
                                       false, false, false]</rosparam>

        <param name="imu0_differential" value="false"/>
        <param name="imu0_relative" value="true"/>
        <param name="imu0_remove_gravitational_acceleration" value="true"/>
        <param name="gravitational_acceleration" value="9.852"/>

        <!-- DR from thrusters -->
        <param name="odom0" value="/sam_thrust_dr"/>
        <rosparam param="odom0_config">[false,  false,  false,
		                                false, false, false,
		                                true,  false, false,
		                                false, false, false,
		                                false, false, false]</rosparam>

        <param name="odom0_differential" value="false"/>
        <param name="odom0_relative" value="false"/>

        <!-- Node params -->
        <param name="publish_tf" value="true"/>
        <param name="publish_acceleration" value="false"/>

        <!-- Advanced node params -->
        <param name="use_control" value="false"/>
        <param name="stamped_control" value="false"/>

        <!-- Outputs -->
        <remap from="odometry/filtered" to="odometry/filtered/sam"/>
        <remap from="set_pose" to="/initialpose"/> <!--Service to reset initial pose -->
    </node>

    <!-- <include file="$(find smarc_private_auvs)/models/sam_auv/launch/upload_sam_auv.launch"/> -->

</launch>
